### PR TITLE
upgrade the crate jdbc driver to 2.1.1

### PR DESCRIPTION
**Pull request by kovrus**

### CODE REVIEW

**Review by bowenwr — CHANGES_REQUESTED** [review]
I tried to startup the new connector server and run the integration suites against it with our Docker container reference data and could not create a connection:

```2016-12-06 12:43:15.175 ERROR 15322 --- [tp1268959798-17] c.z.c.e.f.p.GenericSQLDataProvider       : Connection factory returned null from createConnection

java.lang.IllegalStateException: Connection factory returned null from createConnection
	at org.apache.commons.dbcp2.PoolableConnectionFactory.makeObject(PoolableConnectionFactory.java:258) ~[commons-dbcp2-2.1.1.jar!/:2.1.1]
	at org.apache.commons.dbcp2.BasicDataSource.validateConnectionFactory(BasicDataSource.java:2304) ~[commons-dbcp2-2.1.1.jar!/:2.1.1]
	at org.apache.commons.dbcp2.BasicDataSource.createPoolableConnectionFactory(BasicDataSource.java:2290) ~[commons-dbcp2-2.1.1.jar!/:2.1.1]
	at org.apache.commons.dbcp2.BasicDataSource.createDataSource(BasicDataSource.java:2039) ~[commons-dbcp2-2.1.1.jar!/:2.1.1]
	at org.apache.commons.dbcp2.BasicDataSource.getConnection(BasicDataSource.java:1533) ~[commons-dbcp2-2.1.1.jar!/:2.1.1]
	at com.zoomdata.connector.example.framework.provider.GenericSQLDataProvider.createConnection(GenericSQLDataProvider.java:464) ~[classes!/:na]
	at com.zoomdata.connector.example.framework.provider.GenericSQLDataProvider.pingSource(GenericSQLDataProvider.java:146) ~[classes!/:na]
	at com.zoomdata.connector.example.framework.service.ZoomdataConnectorService.validateSource(ZoomdataConnectorService.java:57) [classes!/:na]
	at com.zoomdata.gen.edc.ConnectorService$Processor$validateSource.getResult(ConnectorService.java:1097) [edc-api-2.3.1.jar!/:na]
	at com.zoomdata.gen.edc.ConnectorService$Processor$validateSource.getResult(ConnectorService.java:1082) [edc-api-2.3.1.jar!/:na]
	at org.apache.thrift.ProcessFunction.process(ProcessFunction.java:39) [libthrift-0.9.3.jar!/:0.9.3]
	at org.apache.thrift.TBaseProcessor.process(TBaseProcessor.java:39) [libthrift-0.9.3.jar!/:0.9.3]
	at org.apache.thrift.server.TServlet.doPost(TServlet.java:83) [libthrift-0.9.3.jar!/:0.9.3]
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:707) [javax.servlet-api-3.1.0.jar!/:3.1.0]
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790) [javax.servlet-api-3.1.0.jar!/:3.1.0]
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:845) [jetty-servlet-9.3.11.v20160721.jar!/:9.3.11.v20160721]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1689) [jetty-servlet-9.3.11.v20160721.jar!/:9.3.11.v20160721]
	at org.eclipse.jetty.websocket.server.WebSocketUpgradeFilter.doFilter(WebSocketUpgradeFilter.java:225) [websocket-server-9.3.11.v20160721.jar!/:9.3.11.v20160721]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1676) [jetty-servlet-9.3.11.v20160721.jar!/:9.3.11.v20160721]
	at org.springframework.boot.web.filter.ApplicationContextHeaderFilter.doFilterInternal(ApplicationContextHeaderFilter.java:55) [spring-boot-1.4.0.RELEASE.jar!/:1.4.0.RELEASE]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107) [spring-web-4.3.2.RELEASE.jar!/:4.3.2.RELEASE]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1676) [jetty-servlet-9.3.11.v20160721.jar!/:9.3.11.v20160721]
	at org.springframework.boot.actuate.trace.WebRequestTraceFilter.doFilterInternal(WebRequestTraceFilter.java:105) [spring-boot-actuator-1.4.0.RELEASE.jar!/:1.4.0.RELEASE]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107) [spring-web-4.3.2.RELEASE.jar!/:4.3.2.RELEASE]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1676) [jetty-servlet-9.3.11.v20160721.jar!/:9.3.11.v20160721]
	at org.springframework.web.filter.RequestContextFilter.doFilterInternal(RequestContextFilter.java:99) [spring-web-4.3.2.RELEASE.jar!/:4.3.2.RELEASE]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107) [spring-web-4.3.2.RELEASE.jar!/:4.3.2.RELEASE]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1676) [jetty-servlet-9.3.11.v20160721.jar!/:9.3.11.v20160721]
	at org.springframework.web.filter.HttpPutFormContentFilter.doFilterInternal(HttpPutFormContentFilter.java:87) [spring-web-4.3.2.RELEASE.jar!/:4.3.2.RELEASE]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107) [spring-web-4.3.2.RELEASE.jar!/:4.3.2.RELEASE]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1676) [jetty-servlet-9.3.11.v20160721.jar!/:9.3.11.v20160721]
	at org.springframework.web.filter.HiddenHttpMethodFilter.doFilterInternal(HiddenHttpMethodFilter.java:77) [spring-web-4.3.2.RELEASE.jar!/:4.3.2.RELEASE]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107) [spring-web-4.3.2.RELEASE.jar!/:4.3.2.RELEASE]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1676) [jetty-servlet-9.3.11.v20160721.jar!/:9.3.11.v20160721]
	at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:197) [spring-web-4.3.2.RELEASE.jar!/:4.3.2.RELEASE]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107) [spring-web-4.3.2.RELEASE.jar!/:4.3.2.RELEASE]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1676) [jetty-servlet-9.3.11.v20160721.jar!/:9.3.11.v20160721]
	at org.springframework.boot.actuate.autoconfigure.MetricsFilter.doFilterInternal(MetricsFilter.java:107) [spring-boot-actuator-1.4.0.RELEASE.jar!/:1.4.0.RELEASE]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107) [spring-web-4.3.2.RELEASE.jar!/:4.3.2.RELEASE]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1676) [jetty-servlet-9.3.11.v20160721.jar!/:9.3.11.v20160721]
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:581) [jetty-servlet-9.3.11.v20160721.jar!/:9.3.11.v20160721]
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143) [jetty-server-9.3.11.v20160721.jar!/:9.3.11.v20160721]
	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:548) [jetty-security-9.3.11.v20160721.jar!/:9.3.11.v20160721]
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:226) [jetty-server-9.3.11.v20160721.jar!/:9.3.11.v20160721]
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1180) [jetty-server-9.3.11.v20160721.jar!/:9.3.11.v20160721]
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:511) [jetty-servlet-9.3.11.v20160721.jar!/:9.3.11.v20160721]
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:185) [jetty-server-9.3.11.v20160721.jar!/:9.3.11.v20160721]
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1112) [jetty-server-9.3.11.v20160721.jar!/:9.3.11.v20160721]
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141) [jetty-server-9.3.11.v20160721.jar!/:9.3.11.v20160721]
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:134) [jetty-server-9.3.11.v20160721.jar!/:9.3.11.v20160721]
	at org.eclipse.jetty.server.Server.handle(Server.java:524) [jetty-server-9.3.11.v20160721.jar!/:9.3.11.v20160721]
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:319) [jetty-server-9.3.11.v20160721.jar!/:9.3.11.v20160721]
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:253) [jetty-server-9.3.11.v20160721.jar!/:9.3.11.v20160721]
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:273) [jetty-io-9.3.11.v20160721.jar!/:9.3.11.v20160721]
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:95) [jetty-io-9.3.11.v20160721.jar!/:9.3.11.v20160721]
	at org.eclipse.jetty.io.SelectChannelEndPoint$2.run(SelectChannelEndPoint.java:93) [jetty-io-9.3.11.v20160721.jar!/:9.3.11.v20160721]
	at org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.executeProduceConsume(ExecuteProduceConsume.java:303) [jetty-util-9.3.11.v20160721.jar!/:9.3.11.v20160721]
	at org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.produceConsume(ExecuteProduceConsume.java:148) [jetty-util-9.3.11.v20160721.jar!/:9.3.11.v20160721]
	at org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.run(ExecuteProduceConsume.java:136) [jetty-util-9.3.11.v20160721.jar!/:9.3.11.v20160721]
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:671) [jetty-util-9.3.11.v20160721.jar!/:9.3.11.v20160721]
	at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:589) [jetty-util-9.3.11.v20160721.jar!/:9.3.11.v20160721]
	at java.lang.Thread.run(Thread.java:745) [na:1.8.0_77]
```

File: test-server/Dockerfile [hunk 1]
```diff
@@ -35,7 +35,7 @@ RUN apt-get -y install crate
 ADD crate-backup.tar.gz /var/lib/crate
 
 # Expose default Crate ports
-EXPOSE 4200 4300
+EXPOSE 4200 4300 5432-5532
```
Comment (bowenwr) [inline test-server/Dockerfile]: Is it necessary to open the full range of 100 ports?

Comment (bowenwr) [reply test-server/Dockerfile]: The corresponding line in README for `docker run` should also be changed and we should note if you're already running PostgreSQL on port 5432 you'll need to either stop it or bind to a different port.

Comment (kovrus) [reply test-server/Dockerfile]: Yes, it is enough to open expose only 5432.

**Comment by kovrus** [comment]
@bowenwr Can you verify that  CrateDB >= 0.57.0 is running in the docker container.
How does the connection string look like? It must be smth like `crate://localhost:5432/`

**Comment by bowenwr** [comment]
@kovrus Running a single node of Crate 0.57.4, EDC is showing:

```
2016-12-06 15:01:19.200  INFO 15669 --- [tp1268959798-13] c.z.c.e.f.p.GenericSQLDataProvider       : New pool with key SQLConnectionPoolKey(jdbcUrl=crate://localhost:5432, username=null, customParams={})
2016-12-06 15:01:19.203 ERROR 15669 --- [tp1268959798-13] c.z.c.e.f.p.GenericSQLDataProvider       : Connection factory returned null from createConnection
```

I will grab a SQL client and try to verify the Crate connection independently later today.

**Comment by kovrus** [comment]
Thanks @bowenwr

 `crate://localhost:5432` must be `crate://localhost:5432/`

**Comment by bowenwr** [comment]
@kovrus Thanks for the help on the JDBC URL. I was able to connect, but running our suite of integration tests the majority fail.

```
29 OK, 145 FAIL, 0 SKIP
Total score: 16.7%
```

We'll need to work through the connector to see what changes we need to make.

**Comment by bowenwr** [comment]
Speaking with @kovrus, it looks like our tests are failing because Crate now returns timestamps as Unix epoch strings while the integration test suite is looking for ISO date strings.

We have a fix that is pending for the test suite to handle this exact case, I will discuss with the team.

**Comment by bowenwr** [comment]
Hi @kovrus, this should be handled by https://github.com/Zoomdata/edc-cratedb/pull/14 if you approve of the changes there.

**Comment by bowenwr** [comment]
Done in scope of dev/upgrade-crate-and-driver, already merged. Thanks for the contribution!

**Pull request closed without merging**